File: backend/schemas/product.py
from datetime import datetime
from typing import List

from pydantic import BaseModel
from schemas.category import CategoryNameResponse
from schemas.color import ColorResponse
from schemas.country import CountryResponse
from schemas.material import MaterialResponse
from schemas.product_attribute import ProductAttributeResponse
from schemas.product_image import ProductImageResponse

class ProductResponse(BaseModel):
    id: int
    name: str
    description: str | None
    category_id: int | None
    pack_quantity: int | None
    quantity: int | None
    price: float | None
    weight: str | None
    color_id: int | None
    material_id: int | None
    country_id: int | None
    purpose: str | None
    created_at: datetime | None
    updated_at: datetime | None
    category: CategoryNameResponse | None = None
    color: ColorResponse | None = None
    material: MaterialResponse | None = None
    country: CountryResponse | None = None
    product_attributes: List[ProductAttributeResponse] = []
    images: List[ProductImageResponse] = []

    class Config:
        from_attributes = True


class ProductCreate(BaseModel):
    name: str
    description: str | None = None
    category_id: int | None = None
    pack_quantity: int | None = None
    quantity: int | None = None
    price: float | None = None
    weight: str | None = None
    color_id: int | None = None
    material_id: int | None = None
    country_id: int | None = None
    purpose: str | None = None


class ProductUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    category_id: int | None = None
    pack_quantity: int | None = None
    quantity: int | None = None
    price: float | None = None
    weight: str | None = None
    color_id: int | None = None
    material_id: int | None = None
    country_id: int | None = None
    purpose: str | None = None

File: frontend/src/types/product.ts
export interface ProductRelation {
  id: number
  name: string
}

export interface ProductColor extends ProductRelation {
  code: string | null
}

export interface ProductImage {
  id: number
  product_id: number
  url: string | null
}

export interface ProductAttributeDefinition {
  id: number
  name: string
}

export interface ProductAttribute {
  product_id: number
  attribute_id: number
  value_text: string | null
  attribute: ProductAttributeDefinition | null
}

export interface Product {
  id: number
  name: string
  description: string | null
  category_id: number | null
  pack_quantity: number | null
  quantity: number | null
  price: number | null
  weight: string | null
  color_id: number | null
  material_id: number | null
  country_id: number | null
  purpose: string | null
  created_at: string | null
  updated_at: string | null
  category: ProductRelation | null
  color: ProductColor | null
  material: ProductRelation | null
  country: ProductRelation | null
  product_attributes: ProductAttribute[]
  images: ProductImage[]
}
